### ISSUE TITLE
Libre2 gen2

### ISSUE THREAD

**Issue by Anton-1408** [gui-dos/DiaBLE#21]
Hi, do I understand correctly that the libre2 Gen2 sensor type sensor work is not finished yet? If not, can I know when the work with this sensor will be added?

**Comment by gui-dos** [comment]
The native algorithms to decrypt the NFC/BLE data were unveiled originally by the Russian FreeAPS X team: I fear that none of the current Trio/iAPS and Android’s xDrip+ guys have the capabilities and/or the interest to reverse-engineer the Gen2 native methods `p1()` and `p2()`…

Juggluco’s technique to redistribute and inject into the original LibreLink obfuscated _libdataprocessing.so_ library is not technically feasible on iOS, even because it is statically linked into the main executable.

**Comment by Anton-1408** [comment]
and for libre 2 the application works, as I understood from the code?

**Comment by gui-dos** [comment]
Yep: I was the first one to decide to publish the full Swift code for the European models that I ported from TypeScript after it was partially unveiled by https://github.com/ivalkou/LibreTools since I discovered that the decrypting keys were stored in the partial 8KB dump of the firmware that you can get after sending the custom NFC command `A1 1A`.

You can spot there even the offset table of the seven trend values transmitted every minute (`[0, 2, 4, 6, 7, 12, 15]`) while the other ports like DiaBox's and Android's xDrip+ ones just extracted the latest value B-)

**Comment by Anton-1408** [comment]
have you tried using libre 2 with the bubble nano sensor? https://github.com/JohanDegraeve/xdripswift/pull/563

**Comment by gui-dos** [comment]
No: I was just renewed my annual prescription for the Libre 3 and for the past two years I have been focusing on it but I left the support for the Bubble/MiaoMiao transmitters in the playground: the modifications would be minimal to support the Nano.

May I ask you whether you own the newer Libre 2 specific for the Russian market?

I am reading that the current workaround to scrape online the `libreview.io` servers doesn't work unless you change the domain into `.ru`: with my iPad playground it should be just a matter of a minute to test it ;-)

If you scan the newer sensor kind, DiaBLE should display "Libre 2 Gen2" as its type and a "Eastern / Rest of World" as its region. Just a first step that other apps haven’t even begun to implement. B-)

**Comment by Anton-1408** [comment]
I'm from Russia. We have a libre 2 gen 2 sensor.

**Comment by Anton-1408** [comment]
Thank you for the answers

**Comment by gui-dos** [comment]
I noticed that there is a slight variation between the `PatchInfo`s of the recent Libre 2 sensors introduced in Russia and in Latin America:

https://github.com/gui-dos/DiaBLE/blob/3932dc36614627d4ee53114931af60f1c87f6cdb/DiaBLE/Libre.swift#L68-L69

There is just an increment in the sixth nibble from `9` to `A` but it could be a substantial bump since the Latin Plus models last one more day and should be certified for integration with the last generation pumps for hybrid looping with commercial solutions.

**Comment by kenkit** [comment]
> No: I was just renewed my annual prescription for the Libre 3 and for the past two years I have been focusing on it but I left the support for the Bubble/MiaoMiao transmitters in the playground: the modifications would be minimal to support the Nano.

Hi, any updates on the libre 3 or 3+ it would be awesome to be able to have access to this devices ?

**Comment by gui-dos** [comment]
I’ve identified the enums and struct member definitions in the decompiled Swift executables but, unfortunately, Juggluco’s developer lacks experience in iOS reverse-engineering and nobody else appears interested in documenting the underlying protocols openly on GitHub anymore.
